File: stripe_terminal/android/src/main/kotlin/com/stripe_terminal/api/ToApiExtensions.kt
package com.stripe_terminal.api

import com.stripe_terminal.toHashMap
import com.stripe.stripeterminal.external.models.*
import java.util.Calendar

fun TerminalException.toApi(): TerminalExceptionApi {
    return  TerminalExceptionApi(
        code = errorCode.name,
        message = errorMessage,
        details = stackTraceToString()
    )
}

fun Reader.toApi(): ReaderApi {
    return ReaderApi(
        locationStatus = locationStatus.toApi(),
        batteryLevel = batteryLevel?.toDouble() ?: -1.0,
        deviceType = deviceType.toApi(),
        simulated = isSimulated,
        availableUpdate = availableUpdate?.hasFirmwareUpdate ?: false,
        locationId = location?.id,
        label = label,
        serialNumber = serialNumber!!,
    )
}

fun LocationStatus.toApi(): LocationStatusApi {
    return when (this) {
        LocationStatus.UNKNOWN -> LocationStatusApi.UNKNOWN
        LocationStatus.SET -> LocationStatusApi.SET
        LocationStatus.NOT_SET -> LocationStatusApi.NOT_SET
    }
}

fun DeviceType.toApi(): DeviceTypeApi {
    return when (this) {
        DeviceType.CHIPPER_1X -> DeviceTypeApi.CHIPPER1_X
        DeviceType.CHIPPER_2X -> DeviceTypeApi.CHIPPER2_X
        DeviceType.STRIPE_M2 -> DeviceTypeApi.STRIPE_M2
        DeviceType.COTS_DEVICE -> DeviceTypeApi.COTS_DEVICE
        DeviceType.VERIFONE_P400 -> DeviceTypeApi.VERIFONE_P400
        DeviceType.WISECUBE -> DeviceTypeApi.WISE_CUBE
        DeviceType.WISEPAD_3 -> DeviceTypeApi.WISEPAD3
        DeviceType.WISEPAD_3S -> DeviceTypeApi.WISEPAD3S
        DeviceType.WISEPOS_E -> DeviceTypeApi.WISEPOS_E
        DeviceType.WISEPOS_E_DEVKIT -> DeviceTypeApi.WISEPOS_E_DEVKIT
        DeviceType.ETNA -> DeviceTypeApi.ETNA
        DeviceType.STRIPE_S700 -> DeviceTypeApi.STRIPE_S700
        DeviceType.STRIPE_S700_DEVKIT -> DeviceTypeApi.STRIPE_S700_DEVKIT
        DeviceType.UNKNOWN -> DeviceTypeApi.UNKNOWN
    }
}

fun ConnectionStatus.toApi(): ConnectionStatusApi {
    return when (this) {
        ConnectionStatus.NOT_CONNECTED -> ConnectionStatusApi.NOT_CONNECTED
        ConnectionStatus.CONNECTING -> ConnectionStatusApi.CONNECTING
        ConnectionStatus.CONNECTED -> ConnectionStatusApi.CONNECTED
    }
}

fun ReaderSoftwareUpdate.toApi(): ReaderSoftwareUpdateApi {
    return ReaderSoftwareUpdateApi(
        components = components.map { it.toApi() },
        hasConfigUpdate = hasConfigUpdate,
        hasFirmwareUpdate = hasFirmwareUpdate,
        hasIncrementalUpdate = hasIncrementalUpdate,
        hasKeyUpdate = hasKeyUpdate,
        keyProfileName = keyProfileName,
        onlyInstallRequiredUpdates = onlyInstallRequiredUpdates,
        requiredAt = requiredAt.time,
        settingsVersion = settingsVersion,
        timeEstimate = timeEstimate.toApi(),
        version = version,
    )
}

fun ReaderSoftwareUpdate.UpdateComponent.toApi(): UpdateComponentApi {
    return when (this) {
        ReaderSoftwareUpdate.UpdateComponent.INCREMENTAL -> UpdateComponentApi.INCREMENTAL
        ReaderSoftwareUpdate.UpdateComponent.FIRMWARE -> UpdateComponentApi.FIRMWARE
        ReaderSoftwareUpdate.UpdateComponent.CONFIG -> UpdateComponentApi.CONFIG
        ReaderSoftwareUpdate.UpdateComponent.KEYS -> UpdateComponentApi.KEYS
    }
}

fun ReaderSoftwareUpdate.UpdateTimeEstimate.toApi(): UpdateTimeEstimateApi {
    return when (this) {
        ReaderSoftwareUpdate.UpdateTimeEstimate.LESS_THAN_ONE_MINUTE -> UpdateTimeEstimateApi.LESS_THAN_ONE_MINUTE
        ReaderSoftwareUpdate.UpdateTimeEstimate.ONE_TO_TWO_MINUTES -> UpdateTimeEstimateApi.ONE_TO_TWO_MINUTES
        ReaderSoftwareUpdate.UpdateTimeEstimate.TWO_TO_FIVE_MINUTES -> UpdateTimeEstimateApi.TWO_TO_FIVE_MINUTES
        ReaderSoftwareUpdate.UpdateTimeEstimate.FIVE_TO_FIFTEEN_MINUTES -> UpdateTimeEstimateApi.FIVE_TO_FIFTEEN_MINUTES

    }
}

fun PaymentMethod.toApi(): PaymentMethodApi {
    return PaymentMethodApi(
        cardDetails = cardDetails?.toApi(),
//        cardPresentDetails
        customer = customer,
        id = id,
        // interacPresentDetails
        livemode = livemode,
        metadata = metadata?.toHashMap(),
    )
}

fun CardDetails.toApi(): CardDetailsApi {
    return CardDetailsApi(
        brand = brand,
        country = country,
        expMonth = expMonth.toLong(),
        expYear = expYear.toLong(),
        fingerprint = fingerprint,
        funding = funding,
//        generatedFrom
        last4 = last4,
    )
}

fun PaymentIntent.toApi(): PaymentIntentApi {
    return PaymentIntentApi(
        amount = amount.toDouble(),
        amountCapturable = amountCapturable.toDouble(),
//         amountDetails
        amountReceived = amountReceived.toDouble(),
//         amountTip
        application = application,
        applicationFeeAmount = applicationFeeAmount.toDouble(),
        canceledAt = canceledAt,
        cancellationReason = cancellationReason,
        captureMethod = captureMethod,
        clientSecret = clientSecret,
        confirmationMethod = confirmationMethod,
        created = created,
        currency = currency,
        description = description,
        id = id,
        invoice = invoice,
//         lastPaymentError
        livemode = livemode,
        metadata = metadata?.toHashMap(),
//         offlineBehavior
        onBehalfOf = onBehalfOf,
//         paymentMethod
//         paymentMethodData
        paymentMethodId = paymentMethodId,
//         paymentMethodOptions
//         paymentMethodUnion
        receiptEmail = receiptEmail,
        review = review,
        setupFutureUsage = setupFutureUsage,
//         statementDescriptor
//         statementDescriptorSuffix
        status = status?.toApi(),
//        stripeAccountId
        transferGroup = transferGroup,
        customer = customer,
    )
}

fun PaymentIntentStatus.toApi(): PaymentIntentStatusApi {
    return when (this) {
        PaymentIntentStatus.CANCELED -> PaymentIntentStatusApi.CANCELED
        PaymentIntentStatus.PROCESSING -> PaymentIntentStatusApi.PROCESSING
        PaymentIntentStatus.REQUIRES_CAPTURE -> PaymentIntentStatusApi.REQUIRES_CAPTURE
        PaymentIntentStatus.REQUIRES_CONFIRMATION -> PaymentIntentStatusApi.REQUIRES_CONFIRMATION
        PaymentIntentStatus.REQUIRES_PAYMENT_METHOD -> PaymentIntentStatusApi.REQUIRES_PAYMENT_METHOD
        PaymentIntentStatus.SUCCEEDED -> PaymentIntentStatusApi.SUCCEEDED
    }
}

fun Location.toApi(): LocationApi {
    return LocationApi(
        address = address?.toApi(),
        displayName = displayName,
        id = id,
        livemode = livemode,
        metadata = metadata?.toHashMap(),
    )
}

fun Address.toApi(): AddressApi {
    return AddressApi(
        city = city,
        country = country,
        line1 = line1,
        line2 = line2,
        postalCode = postalCode,
        state = state,
    )
}

fun PaymentStatus.toApi(): PaymentStatusApi {
    return when (this) {
        PaymentStatus.NOT_READY -> PaymentStatusApi.NOT_READY
        PaymentStatus.READY -> PaymentStatusApi.READY
        PaymentStatus.WAITING_FOR_INPUT -> PaymentStatusApi.WAITING_FOR_INPUT
        PaymentStatus.PROCESSING -> PaymentStatusApi.PROCESSING
    }
}

File: stripe_terminal/ios/Classes/ToApiExtensions.swift
import Foundation
import StripeTerminal

extension Location {
    func toApi() -> LocationApi {
        return LocationApi(
            address: address?.toApi(),
            displayName: displayName,
            id: stripeId,
            livemode: livemode,
            metadata: metadata
        )
    }
}

extension Reader {
    func toApi() -> StripeReaderApi {
        return StripeReaderApi(
            locationStatus: locationStatus.toApi(),
            batteryLevel: batteryLevel?.doubleValue ?? -1.0,
            deviceType: deviceType.toApi(),
            simulated: simulated,
            availableUpdate: availableUpdate != nil,
            locationId: locationId,
            serialNumber: serialNumber,
            label: label
        )
    }
}

extension PaymentIntentStatus {
    func toApi() -> PaymentIntentStatusApi {
        switch self {
        case .requiresPaymentMethod:
            return .requiresPaymentMethod
        case .requiresConfirmation:
            return .requiresConfirmation
        case .requiresCapture:
            return .requiresCapture
        case .processing:
            return .processing
        case .canceled:
            return .canceled
        case .succeeded:
            return .succeeded
        @unknown default:
            fatalError("WTF")
        }
    }
}

extension Address {
    func toApi() -> AddressApi? {
        return AddressApi(
            city: city,
            country: country,
            line1: line1,
            line2: line2,
            postalCode: postalCode,
            state: state
        )
    }
}

extension ConnectionStatus {
    func toApi() -> ConnectionStatusApi {
        switch self {
        case .notConnected:
            return .notConnected
        case .connected:
            return .connected
        case .connecting:
            return .connecting
        @unknown default:
            fatalError("WTF")
        }
    }
}

extension DiscoveryMethod {
    func toApi() -> DiscoveryMethodApi {
        switch self {
        case .bluetoothScan:
            return .bluetoothScan
        case .bluetoothProximity:
            fatalError("Not supported DiscoveryMethod.bluetoothProximity")
        case .internet:
            return .internet
        case .localMobile:
            return .localMobile
        @unknown default:
            fatalError("WTF")
        }
    }
}

extension LocationStatus {
    func toApi() -> LocationStatusApi {
        switch self {
        case .unknown:
            return .unknown
        case .set:
            return .set
        case .notSet:
            return .notSet
        @unknown default:
            fatalError("WTF")
        }
    }
}

extension DeviceType {
    func toApi() -> DeviceTypeApi {
        switch self {
        case .chipper2X:
            return .chipper2X
        case .verifoneP400:
            return .verifoneP400
        case .wisePad3:
            return .wisepad3
        case .stripeM2:
            return .stripeM2
        case .wisePosE:
            return .wiseposE
        case .wisePosEDevKit:
            return .wiseposEDevkit
        case .etna:
            return .etna
        case .chipper1X:
            return .chipper1X
        case .wiseCube:
            return .wiseCube
        case .stripeS700:
            return .stripeS700
        case .stripeS700DevKit:
            return .stripeS700Devkit
        case .appleBuiltIn:
            fatalError("Not supported DeviceType.appleBuiltIn")
        @unknown default:
            fatalError("WTF")
        }
    }
}

extension PaymentStatus {
    func toApi() -> PaymentStatusApi {
        switch self {
        case .notReady:
            return .notReady
        case .ready:
            return .ready
        case .waitingForInput:
            return .waitingForInput
        case .processing:
            return .processing
        @unknown default:
            fatalError("WTF")
        }
    }
}

extension PaymentMethod {
    func toApi() -> StripePaymentMethodApi {
        return StripePaymentMethodApi(
            id: stripeId,
            cardDetails: card?.toApi(),
            customer: customer,
            livemode: true,
            metadata: metadata
        )
    }
}

extension CardDetails {
    func toApi() -> CardDetailsApi {
        return CardDetailsApi(
            brand: brand.toApi(),
            country: country,
            expMonth: expMonth,
            expYear: expYear,
            fingerprint: fingerprint,
            funding: funding.toApi(),
            last4: last4
        )
    }
}

extension CardBrand {
    func toApi() -> String {
        switch self {
        case .amex:
            return "amex"
        case .dinersClub:
            return "diners"
        case .discover:
            return "discover"
        case .JCB:
            return "jcb"
        case .masterCard:
            return "mastercard"
        case .unionPay:
            return "unionpay"
        case .visa:
            return "visa"
        case .unknown:
            return "unknown"
        case .interac:
            return "interac"
        case .eftposAu:
            return "eftposau"
        @unknown default:
            fatalError("WTF")
        }
    }
}

extension CardFundingType {
    func toApi() -> String {
        switch self {
        case .credit:
            return "credit"
        case .debit:
            return "debit"
        case .prepaid:
            return "prepaid"
        case .other:
            return "unknown"
        @unknown default:
            fatalError("WTF")
        }
    }
}

extension PaymentIntent {
    func toApi() -> StripePaymentIntentApi {
        return StripePaymentIntentApi(
            id: stripeId,
            amount: Double(amount),
            amountCapturable: 0,
            amountReceived: 0,
            application: nil,
            applicationFeeAmount: nil,
            captureMethod: captureMethod.toApi(),
            cancellationReason: nil,
            canceledAt: nil,
            clientSecret: nil,
            confirmationMethod: nil,
            created: created,
            currency: currency,
            customer: nil,
            description: description,
            invoice: nil,
            livemode: false,
            metadata: metadata,
            onBehalfOf: nil,
            paymentMethodId: nil,
            status: status.toApi(),
            review: nil,
            receiptEmail: nil,
            setupFutureUsage: nil,
            transferGroup: nil
        )
    }
}

extension CaptureMethod {
    func toApi() -> String {
        switch self {
        case .manual:
            return "manual"
        case .automatic:
            return "automatic"
        @unknown default:
            fatalError("WTF")
        }
    }
}

extension NSError {
    func toApi() -> String {
        let code = AppleBuiltInReaderErrorCode(rawValue: self.code)
        switch code! {
        case .unknown:
            return "unknown"
        case .unexpectedNil:
            return "unexpectedNil"
        case .invalidTransactionType:
            return "invalidTransactionType"
        case .passcodeDisabled:
            return "passcodeDisabled"
        case .notAllowed:
            return "notAllowed"
        case .backgroundRequestNotAllowed:
            return "backgroundRequestNotAllowed"
        case .unsupported:
            return "unsupported"
        case .osVersionNotSupported:
            return StripeTerminalExceptionCodeApi.localMobileUnsupportedAndroidVersion.rawValue // TODO: fix my name
        case .modelNotSupported:
            return StripeTerminalExceptionCodeApi.unsupportedOperation.rawValue
        case .networkError:
            return StripeTerminalExceptionCodeApi.stripeApiConnectionError.rawValue
        case .networkAuthenticationError:
            return StripeTerminalExceptionCodeApi.stripeApiError.rawValue
        case .serviceConnectionError:
            return "serviceConnectionError"
        case .notReady:
            return StripeTerminalExceptionCodeApi.sessionExpired.rawValue
        case .emptyReaderToken:
            return "emptyReaderToken"
        case .invalidReaderToken:
            return "invalidReaderToken"
        case .prepareFailed:
            return "prepareFailed"
        case .prepareExpired:
            return "prepareExpired"
        case .tokenExpired:
            return "tokenExpired"
        case .deviceBanned:
            return "deviceBanned"
        case .readerMemoryFull:
            return "readerMemoryFull"
        case .readerBusy:
            return StripeTerminalExceptionCodeApi.readerBusy.rawValue
        case .accountNotLinked:
            return "accountNotLinked"
        case .accountLinkingFailed:
            return "accountLinkingFailed"
        case .accountLinkingRequiresiCloudSignIn:
            return "accountLinkingRequiresiCloudSignIn"
        case .accountLinkingCancelled:
            return "accountLinkingCancelled"
        case .merchantBlocked:
            return "merchantBlocked"
        case .invalidMerchant:
            return "invalidMerchant"
        case .readNotAllowed:
            return "readNotAllowed"
        case .readFromBackgroundError:
            return "readFromBackgroundError"
        case .readerServiceConnectionError:
            return "readerServiceConnectionError"
        case .readerServiceError:
            return "readerServiceError"
        case .noReaderSession:
            return "noReaderSession"
        case .readerSessionExpired:
            return "readerSessionExpired"
        case .readerTokenExpired:
            return "readerTokenExpired"
        case .readerSessionNetworkError:
            return "readerSessionNetworkError"
        case .readerSessionAuthenticationError:
            return "readerSessionAuthenticationError"
        case .readerSessionBusy:
            return StripeTerminalExceptionCodeApi.readerBusy.rawValue
        case .readCancelled:
            return "readCancelled"
        case .invalidAmount:
            return "invalidAmount"
        case .invalidCurrency:
            return "invalidCurrency"
        case .nfcDisabled:
            return StripeTerminalExceptionCodeApi.localMobileNfcDisabled.rawValue
        case .readNotAllowedDuringCall:
            return "readNotAllowedDuringCall"
        case .cardReadFailed:
            return "cardReadFailed"
        case .paymentReadFailed:
            return "paymentReadFailed"
        case .paymentCardDeclined:
            return "paymentCardDeclined"
        case .pinEntryFailed:
            return "pinEntryFailed"
        case .pinTokenInvalid:
            return "pinTokenInvalid"
        case .pinEntryTimeout:
            return "pinEntryTimeout"
        case .pinCancelled:
            return "pinCancelled"
        case .pinNotAllowed:
            return "pinNotAllowed"
        default:
            return ""
        }
    }
}
